lib/command/src/Obelisk/Command/Preprocessor.hs

```diff
@@ -0,0 +1,92 @@
+import System.Environment
+import System.IO
+import System.FilePath
+
+import Obelisk.App
+import Obelisk.Command (mkObeliskConfig)
+import Obelisk.Command.Run
+
+main :: IO ()
+main = do
+  -- The command line argumens are passed in via ghci.
+  -- ghci dictates the first three options and meanings
+  -- TODO: perform some validation and print a nice error/help message, if we don't see something plausible.
+  -- We do have control over the remaining arguments, but they must be the same for all files.
+  -- We assume all the remaining arguments passed in are paths to cabal or hpack package specifications.
+
+  args@(origPath:inPath:outPath:packagePaths) <- getArgs
+
+  outFile <- openFile outPath WriteMode
+  let hPutTextBuilder h = BU.hPutBuilder h . TL.encodeUtf8Builder . TL.toLazyText
+
+  putStr "--------------------------------------------------------------------------------\n"
+  print args
+  putStr "--------------------------------------------------------------------------------\n"
+
+  -- Thus we must select among the packagePaths for the file we are going to parse.
+
+  let takeDirs = takeWhile hasTrailingPathSeparator
+      packageDirs = sortOn (negate . length . takeDirs) $ map (splitPath . normalise) packagePaths
+      origDir = splitPath $ normalise $ origPath
+      matches = [ joinPath d | d <- packageDirs, takeDirs d `isPrefixOf` origDir ]
+
+  -- So the first element of matches is going to be the deepest path to a package spec that contains
+  -- our file as a subdirectory.
+
+  config <- mkObeliskConfig
+
+  case matches of
+    [] -> hPutTextBuilder outFile (lineNumberPragma origPath) -- TODO: probably should produce a warning
+    (packagePath:_) -> do
+       runObelisk config (parseCabalPackage packagePath) >>= \case
+         Just packageInfo -> do
+           hPutTextBuilder outFile (generateHeader origPath packageInfo)
+
+  BL.readFile inPath >>= BL.hPut outFile
+  hClose outFile
+
+
+ -- I'm pretty sure there's a certain amount of oversimplification in CabalPackageInfo, so I doubt this is fully robust.
+
+generateHeader :: FilePath -> CabalPackageInfo -> TL.Builder
+generateHeader origPath packageInfo =
+    hsExtensions <> ghcOptions <> lineNumberPragma origPath
+  where
+    hsExtensions =
+      if not (null extList)
+      then TL.fromText "{-# LANGUAGE "
+        <> mconcat (intersperse (TL.fromText ", ") extList)
+        <> TL.fromText " #-}\n"
+      else mempty
+    extList = concatMap showExt (_cabalPackageInfo_defaultExtensions packageInfo)
+    showExt = \case
+      EnableExtension ext -> [TL.fromString (show ext)]
+      DisableExtension _ -> []
+      UnknownExtension ext -> [TL.fromString ext]
+
+    ghcOptions =
+      if not (null optList)
+      then TL.fromText "{-# OPTIONS_GHC "
+        <> mconcat (intersperse (TL.fromText " ") optList)
+        <> TL.fromText " #-}\n"
+      else mempty
+    optList = map TL.fromString $ fromMaybe [] $ lookup GHC (_cabalPackageInfo_compilerOptions packageInfo)
```

Comment:
I wonder what happens when there are conditionals in the cabal file... we should test that.